# Python : manipulation de strings 

## Code morse
Jm.Torres, Janvier 2023

### Nous allons programmer un convertisseur : texte <-> code morse 

Le code morse est défini par une séquence de "points" et de "traits" pour chaque lettre, comme le montre la figure. Ainsi la fameux message "S.O.S" (Save Our Souls) vaut : `... --- ...` 

On utilise aussi "ti" pour parler d'un point et "ta" pour un trait (alors SOS se dit : `ti-ti-ti--ta-ta-ta--ti-ti-ti`, notez les `--` pour figurer la séparation des lettres, et les `-` pour séparer les signes)

<img src="morse.png" alt="drawing" width="350"/>

Pour ce qui nous concerne, nous avons des 0 et des 1, et nous allons coder comme ceci :

- "1" pour un "ti" (point),
- "11" pour un "ta" (trait), 
- "0" pour séparer deux signes ("ta" ou "ti"),
- "00" pour séparer deux lettres,
- "000" pour séparer deux mots. 

Par exemple, la chaine `code morse` devient :  `-.-. --- -.. .    -- --- .-. ... .`


Et avec notre convention avec les 0 et les 1 : `110101101001101101100110101001000110110011011011001011010010101001000`




On donne le dictionnaire M2A (morse to alphabet) et A2M (alphabet to morse) : 


In [1]:
M2A = { '.-':'a', '-...':'b', '-.-.':'c', '-..':'d', '.':'e', '..-.':'f', '--.':'g', 
     '....':'h', '..':'i', '.---':'j', '-.-':'k', '.-..':'l', '--':'m', '-.':'n',
     '---':'o', '.--.':'p', '--.-':'q', '.-.':'r', '...':'s', '-':'t', '..-':'u',
     '...-':'v', '.--':'w', '-..-':'x', '-.--':'y', '--..':'z', '.----':'1', 
     '..---':'2', '...--':'3', '....-':'4', '.....':'5', '-....':'6', '--...':'7',
     '---..':'8', '----.':'9', '-----':'0', '--..--':',', '.-.-.-':'.', '..--..':'?',
     '-..-.':'/', '-....-':'-', '--..--':',', '---...':':', '-.--.':'(', '-.--.-':')',
     '..-..':'é','.--.-':'à', '-.-.--':'!'} 

A2M = { 'a':'.-', 'b':'-...', 'c':'-.-.', 'd':'-..', 'e':'.', 
        'f':'..-.', 'g':'--.', 'h':'....', 'i':'..', 'j':'.---', 'k':'-.-', 
        'l':'.-..', 'm':'--', 'n':'-.', 'o':'---', 'p':'.--.', 'q':'--.-', 
        'r':'.-.', 's':'...', 't':'-', 'u':'..-', 'v':'...-', 'w':'.--', 
        'x':'-..-', 'y':'-.--', 'z':'--..', '1':'.----', '2':'..---', 
        '3':'...--', '4':'....-', '5':'.....', '6':'-....', '7':'--...',
        '8':'---..', '9':'----.', '0':'-----', ', ':'--..--', '.':'.-.-.-', 
        '?':'..--..', '/':'-..-.', '-':'-....-', ',':'--..--',':':'---...',
        '(':'-.--.', ')':'-.--.-','é':'..-..','à':'.--.-', '!':'-.-.--'} 

# 1. Décoder du morse vers un texte en clair
### On commence par "séparer" les "mots", c'est à dire les parties du message morse qui sont séparées par '000'. 
On peut utiliser la méthode `split` : appliquée à une chaine de caractères, elle renvoie la liste des chaines séparées, par défaut split utilise le caractère blanc/espace epour sépraere les mots, mais on peut spécifier le séparateur que l'on veut, exemples:

In [2]:
# exemples d'utilisation de '.split()'
s = "bonjour tout le monde"
mots = s.split()
print(s, "->", mots)

s = "pommes,poires,fraises des bois,framboises"
fruits = s.split(",")
print(s, "->", fruits)

bonjour tout le monde -> ['bonjour', 'tout', 'le', 'monde']
pommes,poires,fraises des bois,framboises -> ['pommes', 'poires', 'fraises des bois', 'framboises']


In [145]:
message = '110110010010110100110101101001010001011011010011011011001010110010110100010101011001101101100110010110100100010110011001100100110100110010100110110110011010011011010101101100011010110100100110001011001100100101101010010100100101101000100101010011000110110010110010100110100110010011010010110011010011000110010010110100110110010100110100101011010100110110101011011000110101010010110100101100101010110011011011000110101101011011'

In [146]:
# afficher la liste des "mots" du message (séparés par 000)
mes = message.split('000')
print(message, "->",mes)

110110010010110100110101101001010001011011010011011011001010110010110100010101011001101101100110010110100100010110011001100100110100110010100110110110011010011011010101101100011010110100100110001011001100100101101010010100100101101000100101010011000110110010110010100110100110010011010010110011010011000110010010110100110110010100110100101011010100110110101011011000110101010010110100101100101010110011011011000110101101011011 -> ['11011001001011010011010110100101', '10110110100110110110010101100101101', '101010110011011011001100101101001', '101100110011001001101001100101001101101100110100110110101011011', '1101011010010011', '1011001100100101101010010100100101101', '100101010011', '110110010110010100110100110010011010010110011010011', '110010010110100110110010100110100101011010100110110101011011', '110101010010110100101100101010110011011011', '110101101011011']


In [36]:
# a présent pour chaque mot, afficher chaque lettre (les lettres sont séparées par "00")
# par exemple '101101' est une lettre.
#letters = []
#for m in mes:
#    letter  = m.split('00')
#    letters.append(letter)
#print(letters)

In [147]:
unit=[]
letters = []
for m in mes:
    letter  = m.split('00')
    letters.append(letter)
    letters.append(' ')
    #print(letters)
for l in letters:
    #print('ici',l)
    for u in l:
        #print('u',u)
        un= u.split('0')
        unit.append(un)
print(unit)
print(letters)

[['11', '11'], ['1'], ['1', '11', '1'], ['11', '1', '11', '1'], ['1', '1'], [' '], ['1', '11', '11', '1'], ['11', '11', '11'], ['1', '1', '11'], ['1', '11', '1'], [' '], ['1', '1', '1', '11'], ['11', '11', '11'], ['11'], ['1', '11', '1'], ['1'], [' '], ['1', '11'], ['11'], ['11'], ['1'], ['11', '1'], ['11'], ['1', '1'], ['11', '11', '11'], ['11', '1'], ['11', '11', '1', '1', '11', '11'], [' '], ['11', '1', '11', '1'], ['1'], ['11'], [' '], ['1', '11'], ['11'], ['1'], ['1', '11', '1', '1'], ['1', '1'], ['1'], ['1', '11', '1'], [' '], ['1'], ['1', '1', '1'], ['11'], [' '], ['11', '11'], ['1', '11'], ['1', '1'], ['11', '1'], ['11'], ['1'], ['11', '1'], ['1', '11'], ['11', '1'], ['11'], [' '], ['11'], ['1'], ['1', '11', '1'], ['11', '11'], ['1', '1'], ['11', '1'], ['1', '1', '11', '1', '1'], ['11', '11', '1', '1', '11', '11'], [' '], ['11', '1', '1', '1'], ['1', '11', '1'], ['1', '11'], ['1', '1', '1', '11'], ['11', '11', '11'], [' '], ['11', '1', '11', '1', '11', '11'], [' ']]
[['11011', 

In [148]:
for code in unit:
    for i in range(len(code)):
        if code[i] =='11':
            code[i]='-'
print(unit)

[['-', '-'], ['1'], ['1', '-', '1'], ['-', '1', '-', '1'], ['1', '1'], [' '], ['1', '-', '-', '1'], ['-', '-', '-'], ['1', '1', '-'], ['1', '-', '1'], [' '], ['1', '1', '1', '-'], ['-', '-', '-'], ['-'], ['1', '-', '1'], ['1'], [' '], ['1', '-'], ['-'], ['-'], ['1'], ['-', '1'], ['-'], ['1', '1'], ['-', '-', '-'], ['-', '1'], ['-', '-', '1', '1', '-', '-'], [' '], ['-', '1', '-', '1'], ['1'], ['-'], [' '], ['1', '-'], ['-'], ['1'], ['1', '-', '1', '1'], ['1', '1'], ['1'], ['1', '-', '1'], [' '], ['1'], ['1', '1', '1'], ['-'], [' '], ['-', '-'], ['1', '-'], ['1', '1'], ['-', '1'], ['-'], ['1'], ['-', '1'], ['1', '-'], ['-', '1'], ['-'], [' '], ['-'], ['1'], ['1', '-', '1'], ['-', '-'], ['1', '1'], ['-', '1'], ['1', '1', '-', '1', '1'], ['-', '-', '1', '1', '-', '-'], [' '], ['-', '1', '1', '1'], ['1', '-', '1'], ['1', '-'], ['1', '1', '1', '-'], ['-', '-', '-'], [' '], ['-', '1', '-', '1', '-', '-'], [' ']]


In [149]:
for code in unit:
    for i in range(len(code)):
        if code[i] =='1':
            code[i]='.'
print(unit)

[['-', '-'], ['.'], ['.', '-', '.'], ['-', '.', '-', '.'], ['.', '.'], [' '], ['.', '-', '-', '.'], ['-', '-', '-'], ['.', '.', '-'], ['.', '-', '.'], [' '], ['.', '.', '.', '-'], ['-', '-', '-'], ['-'], ['.', '-', '.'], ['.'], [' '], ['.', '-'], ['-'], ['-'], ['.'], ['-', '.'], ['-'], ['.', '.'], ['-', '-', '-'], ['-', '.'], ['-', '-', '.', '.', '-', '-'], [' '], ['-', '.', '-', '.'], ['.'], ['-'], [' '], ['.', '-'], ['-'], ['.'], ['.', '-', '.', '.'], ['.', '.'], ['.'], ['.', '-', '.'], [' '], ['.'], ['.', '.', '.'], ['-'], [' '], ['-', '-'], ['.', '-'], ['.', '.'], ['-', '.'], ['-'], ['.'], ['-', '.'], ['.', '-'], ['-', '.'], ['-'], [' '], ['-'], ['.'], ['.', '-', '.'], ['-', '-'], ['.', '.'], ['-', '.'], ['.', '.', '-', '.', '.'], ['-', '-', '.', '.', '-', '-'], [' '], ['-', '.', '.', '.'], ['.', '-', '.'], ['.', '-'], ['.', '.', '.', '-'], ['-', '-', '-'], [' '], ['-', '.', '-', '.', '-', '-'], [' ']]


In [150]:
print(unit[5])

[' ']


In [151]:
test=''
for i in range(len(unit)):
    if unit[i]==[' ']:
        unit[i]= ' '
        #print(unit[i])
    else :
        unit[i]= test.join(unit[i])
        unit[i]= M2A.get(unit[i])
    #print(unit[i])
    #tab=[]
    #util = test.join(unit[i])
    #print(util)
    #tab.append(util)
    #print(tab[0])
    #unit[i] = tab[0]
    #unit[i][0]= M2A.get(unit[i])
#print(unit)
print(unit)


['m', 'e', 'r', 'c', 'i', ' ', 'p', 'o', 'u', 'r', ' ', 'v', 'o', 't', 'r', 'e', ' ', 'a', 't', 't', 'e', 'n', 't', 'i', 'o', 'n', ',', ' ', 'c', 'e', 't', ' ', 'a', 't', 'e', 'l', 'i', 'e', 'r', ' ', 'e', 's', 't', ' ', 'm', 'a', 'i', 'n', 't', 'e', 'n', 'a', 'n', 't', ' ', 't', 'e', 'r', 'm', 'i', 'n', 'é', ',', ' ', 'b', 'r', 'a', 'v', 'o', ' ', '!', ' ']


In [84]:
M2A.get(unit[0])

'm'

In [152]:
sentence = test.join(unit)
sentence


'merci pour votre attention, cet atelier est maintenant terminé, bravo ! '

In [21]:
#letters = []
#for m in mes:
#    letter  = m.split('00')
#    letters.append(letter)
#    unit=[]
#    for t in letter:
#        u= t.split('0')
#        unit.append(u)
#print(letters)

[['11011', '1', '101101', '110101101', '101'], ['101101101', '11011011', '101011', '101101'], ['10101011', '11011011', '11', '101101', '1'], ['1011', '11', '11', '1', '1101', '11', '101', '11011011', '1101', '110110101011011'], ['110101101', '1', '11'], ['1011', '11', '1', '10110101', '101', '1', '101101'], ['1', '10101', '11'], ['11011', '1011', '101', '1101', '11', '1', '1101', '1011', '1101', '11'], ['11', '1', '101101', '11011', '101', '1101', '1010110101', '110110101011011'], ['11010101', '101101', '1011', '10101011', '11011011'], ['110101101011011']]


#### Pour finir, on va transformer chaque "lettre" dans la forme ti / ta ( . - ) : 

par exemple `11010101` donne `-...`
Il faut donc séparer chaque lettre en signe (séparateur `0`) et tranformer les `1` en `.` et les `11` en `-`

In [ ]:
# a / obtenir la liste des signes, comme précédemment, avec .split()
mot = '11011001001011010011010110100101'

# par exemple on obtient : 
#['11011', '1', '101101', '110101101', '101']  <- la liste des lettres
#['11', '11']                                  <- les signes pour la première lettre
#['1']                                         <- les signes pour la seconde lettre
#['1', '11', '1']                              <- etc
#['11', '1', '11', '1']                        <- ...
#['1', '1']      

In [ ]:
# b / pour une lettre (par exemple lettre = ['11', '11', '1', '1']) devient lettre = ['-', '-', '.', '.'])
#     en replaçant les '1' par '.' et les '11' par '-'. , on peut utiliser str.replace(old,new)
#     Attention: .replace() ne s'applique pas à une liste emais à ces strings.
lettre = ['11', '11', '1', '1']

In [ ]:
# on va maintenant former la lettre morse : ['-', '-', '.', '.'] devient '--..'
s = ""

In [ ]:
# Il reste à interroger le dictionnaire M2A avec la clef `--..` pour obtenir la lettre de l'alaphabet
l = M2A.get(s)
print(l)

In [ ]:
# il reste à assemble maintenant les lettres dans les mots et les mots dans la phrase 
# en réutilisant tous les codes obtenus ci-dessus.  

In [ ]:
message = '110110010010110100110101101001010001011011010011011011001010110010110100010101011001101101100110010110100100010110011001100100110100110010100110110110011010011011010101101100011010110100100110001011001100100101101010010100100101101000100101010011000110110010110010100110100110010011010010110011010011000110010010110100110110010100110100101011010100110110101011011000110101010010110100101100101010110011011011000110101101011011000'


# 2. Coder un texte en clair vers un message en morse

### La démarche est tout à fait similaire, on sépare le texte en mots, les mots en lettres, on utilise le dictionnaire pour obtenir les lettre en morse,  on transforme en 1 et 11, on assemble avec les 0, 00, 000

In [ ]:
texte = "bonjour tout le monde"

In [ ]:
# séparer le texte en mots

In [ ]:
# séparer le texte en mots et les mots en lettres      

In [ ]:
# séparer le texte en mots et les mots en lettres, chercher les caractères en morse

In [ ]:
# séparer le texte en mots et les mots en lettres, chercher les caractères en morse
# et cumuler